File: src/contest/atcoder/others/CodeFestival/2015/Golf/B/b.py
s=lambda:list(map(int,input().split()))
N,Q=s()
p=[0]*N
g=lambda x:p[x] if p[x] else x
for i in range(Q):
    A,B,C=s()
    if A==1:
        print ['YES','NO'][g(B)==g(C)]
    else :
        if g(B)!=g(C):p[B]=C
        

File: src/contest/atcoder/others/CodeFestival/2015/Golf/B/b.c
n,q,p[100001]={};
a,b,c;
P(x){return p[x]?P(x):x;}
M(x,y){}
main(){
    scanf("%d%d",&n,&q);
    while(q--){
        scanf("%d%d%d",&a,&b,&c);
        if(a)puts(P(b)-P(c)?"YES":"NO");
        else if(P(b)-P(c))?a:p[b]=c;
    }
}
